src/lib.rs

```diff
@@ -48,33 +53,72 @@ pub fn headtail(opts: &Opts) -> Result<()> {
                     }
                 }
             }
-            Err(e) => return Err(e),
         }
     }
 
     // Keep following(?)
-
+    //
+    // To avoid wasted CPU cycles, we can use a file system watcher (e.g.
+    // `inotify(7)` on Linux).
+    //
+    // The `notify` crate provides several optimized file system watchers using
+    // functionality built into operating systems. Should an optimized watcher
+    // not be available, `notify` will default to a polling watcher.
     if opts.follow && opts.filename.is_some() {
-        let sleep_interval = Duration::from_secs_f64(opts.sleep_interval);
         let mut line = String::new();
-        loop {
-            line.clear();
-            match reader.read_line(&mut line) {
-                Ok(0) => {
-                    // This is a busy loop, so add a little sleep to make it less CPU hungry
-                    std::thread::sleep(sleep_interval);
-                }
-                Ok(_) => {
-                    careful_write(&mut writer, &line)?;
-                    let _ = writer.flush();
-                }
-                Err(e) => {
-                    println!("The error is {:?}", e.kind());
-                    return Err(e);
+
+        // Use a channel for synchronization between the watcher and the main
+        // thread.
+        let (tx, rx) = std::sync::mpsc::channel();
+
+        // If using a polling watcher, respect the `--sleep-interval` argument.
+        let sleep_interval = Duration::from_secs_f64(opts.sleep_interval);
+        let config = notify::Config::default().with_poll_interval(sleep_interval);
+
+        // Setup the file watcher
+        let mut watcher = notify::RecommendedWatcher::new(
+            move |res: notify::Result<notify::Event>| match res {
+                Ok(event) if is_modification(&event) => {
+                    line.clear();
+                    match reader.read_line(&mut line) {
+                        Ok(0) => {}
+                        Ok(_) => {
+                            match careful_write(&mut writer, &line) {
+                                Ok(_) => {}
+                                Err(e) => eprintln!("Write error: {:?}", e.kind()),
+                            }
+                            let _ = writer.flush();
+                        }
+                        Err(e) => {
+                            eprintln!("The error is {:?}", e.kind());
+                        }
+                    }
+                    tx.send(()).unwrap();
                 }
-            }
-        }
+                Ok(_) => {}
+                Err(e) => eprintln!("Watcher error: {:?}", e),
+            },
+            config,
+        )?;
+
+        watcher.watch(
```

Comment:
There's a race condition: here we're watching a filename, but we already opened this file and it could have been renamed in the meantime. Then there could be new data flowing into the file and the watcher wouldn't notice because it's watching a different file.

It's possible to write a test for this. The rename would have to happen between opening the file and getting here, which would be tricky timing, but I think by piping stdout to a slow-reading program you can cause the above loop to block in `careful_write` for however long you like (because Unix pipes [have backpressure](https://unix.stackexchange.com/questions/339401/can-writing-to-stdout-place-backpressure-on-a-process)).

Reply:
Oooh, good catch! Would this be as simple as beginning to watch the file _before_ opening it for reading?

Reply:
That's a good idea.

Which one happens first isn't the issue. The issue is the amount of time that can pass between the two.

As long as there's any gap at all, there's technically a race condition. But beginning to watch the file at *very close to the same time* we open it for reading would make the bug extremely unlikely to trigger.

Reply:
...I might be totally wrong about this. There might be a big difference between watching first and opening first. But it depends on what `notify` actually does on the user's platform, and I have no idea. The docs say the behavior ["may be unexpected"](https://docs.rs/notify/latest/notify/trait.Watcher.html#tymethod.watch).

There are some [known issues with `notify` generally](https://docs.rs/notify/latest/notify/index.html#known-problems), e.g. it doesn't work in docker on M1 Macs.

Reply:
I added a TODO to look into this later. I have a suspicious feeling that it might require a big refactor (ie use the watcher for the initial headtail as well).